typescript/sdk/src/core/AbacusCore.ts

```diff
@@ -46,15 +46,41 @@ export class AbacusCore<Chain extends ChainName = ChainName> extends AbacusApp<
     super(contractsMap, multiProvider);
   }
 
-  static fromEnvironment<Env extends CoreEnvironment>(
-    env: Env,
-    multiProvider: MultiProvider<CoreEnvironmentChain<Env>>,
-  ): AbacusCore<CoreEnvironmentChain<Env>> {
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+  static fromEnvironment<
+    Env extends CoreEnvironment,
+    Chain extends ChainName = ChainName,
+  >(env: Env, multiProvider: MultiProvider<Chain>) {
+    const envConfig = environments[env];
+    if (!envConfig) {
+      throw new Error(`No default env config found for ${env}`);
+    }
+
+    const multiProviderChains = multiProvider.chains();
+    const envChains = Object.keys(envConfig);
+    const intersection = multiProviderChains.filter((c) =>
+      envChains.includes(c),
+    );
+
+    if (!intersection.length) {
+      throw new Error(`No chains shared between MultiProvider and env ${env}`);
+    }
+
+    type EnvChain = keyof typeof envConfig;
+    type IntersectionChain = EnvChain & Chain;
+    // Force cast of multiProvider to ensure Core gets correct type later
+    const intersectionProvider =
+      multiProvider as unknown as MultiProvider<IntersectionChain>;
```

Comment:
imo we should just be retyping `AbacusCore` such that this isn't necessary

Reply:
(and `AbacusApp`)

Reply:
Hmm, can you be more specific?